source/material_model/rheology/peierls_creep.cc

```diff
@@ -38,61 +38,153 @@ namespace aspect
+
+        const double s = ( (p.activation_energy + pressure * p.activation_volume) / (constants::gas_constant * temperature)) *
+                         p.glide_parameter_p * p.glide_parameter_q *
+                         std::pow((1. - std::pow(p.fitting_parameter, p.glide_parameter_p)),(p.glide_parameter_q - 1.)) *
+                         std::pow(p.fitting_parameter, p.glide_parameter_p);
+
+        const double stress_term = 0.5 * (p.fitting_parameter * p.peierls_stress) /
+                                   std::pow((p.prefactor * std::pow(p.fitting_parameter * p.peierls_stress,p.stress_exponent)),( 1. / (s + p.stress_exponent)));
+
+        const double arrhenius_term = std::exp( ((p.activation_energy + pressure * p.activation_volume) / (constants::gas_constant * temperature)) *
+                                                (std::pow((1. - std::pow(p.fitting_parameter,p.glide_parameter_p)),p.glide_parameter_q)) /
+                                                (s + p.stress_exponent) );
+
+        const double strain_rate_term = std::pow(strain_rate, (1. / (s + p.stress_exponent)) - 1.);
+
+        return stress_term * arrhenius_term * strain_rate_term;
+      }
+
+
+
+      template <int dim>
+      double
+      PeierlsCreep<dim>::compute_exact_viscosity (const double strain_rate,
+                                                  const double pressure,
+                                                  const double temperature,
+                                                  const unsigned int composition) const
+      {
+        /**
+         * A generalised Peierls creep formulation. The Peierls creep expression
+         * for the strain rate has multiple stress-dependent terms, and cannot be
+         * directly inverted to find an expression for viscosity in terms of
+         * strain rate. For this reason, a Newton Raphson iteration is required,
+         * which can be quite expensive.
+         * The equation for the strain rate is given in
+         * compute_exact_strain_rate_and_derivative.
+         */
+        const PeierlsCreepParameters p = compute_creep_parameters(composition);
+
+        // The generalized Peierls creep flow law cannot be expressed as viscosity in
+        // terms of strain rate, because there are multiple stress-dependent terms
+        // in the strain rate expression.
+        // We use Newton's method to find the second invariant of the stress tensor.
+
+        // Create a starting guess for the stress using
+        // the approximate form of the viscosity expression
+        double viscosity = compute_approximate_viscosity(strain_rate, pressure, temperature, composition);
+        double stress_ii = 2.*viscosity*strain_rate;
+        double strain_rate_residual = 2.*strain_rate_residual_threshold;
+
+        double strain_rate_deriv = 0;
+        unsigned int stress_iteration = 0;
+        while (std::abs(strain_rate_residual) > strain_rate_residual_threshold
```

Comment:
In diffusion_dislocation a fixpoint iteration is used if the Newton iteration does not succeed. Are we likely to run into the same issues here? If so, should we add this option in now? I have no practical experience with this type of rheology, and I will defer to your judgement. Worst case scenario, we add in the fixpoint iterative scheme if we find the Newton method frequently fails. 

However, right now if `stress_max_iteration_number` is reached there is no message to indicate the strain rate residual threshold was not reached.